### PR TITLE
WIP Iceberg integration

### PR DESCRIPTION
**Pull request by margon8**
Add Iceberg support to Metabolic

### CODE REVIEW

**Comment by braislchao** [comment]
Test catalog based readstream in Iceberg reads:
```
data.writeStream
    .format("iceberg")
    .outputMode("append")
    .trigger(Trigger.ProcessingTime(1, TimeUnit.MINUTES))
    .option("checkpointLocation", checkpointPath)
    .toTable("database.table_name")
```

`Queries with streaming sources must be executed with writeStream.start();
local.data_lake.letters
org.apache.spark.sql.AnalysisException: Queries with streaming sources must be executed with writeStream.start();`

**Comment by braislchao** [comment]
For catalog compatibility we can use Iceberg Session Catalog:

> Spark's built-in catalog supports existing v1 and v2 tables tracked in a Hive Metastore. This configures Spark to use Iceberg's SparkSessionCatalog as a wrapper around that session catalog. When a table is not an Iceberg table, the built-in catalog will be used to load it instead.

**Comment by braislchao** [comment]
Testing in glue dev enfironment with new Spark config:

```
    implicit val spark = sparkBuilder
      .appName(s" Metabolic Mapper - $configPath")
      .config("spark.sql.sources.partitionOverwriteMode", "dynamic")
      .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,io.delta.sql.DeltaSparkSessionExtension")
      .config("spark.sql.catalog.spark_catalog", "org.apache.iceberg.spark.SparkSessionCatalog")
      .config("spark.databricks.delta.schema.autoMerge.enabled", "true")
      .config("spark.databricks.delta.optimize.repartition.enabled", "true")
      .config("spark.databricks.delta.vacuum.parallelDelete.enabled", "true")
      .config("spark.sql.catalog.spark_catalog.catalog-impl", "org.apache.iceberg.aws.glue.GlueCatalog")
      .config("spark.sql.catalog.spark_catalog.io-impl", "org.apache.iceberg.aws.s3.S3FileIO")
      .config("spark.sql.catalog.spark_catalog.client.region", "eu-central-1")
      .config("spark.sql.defaultCatalog", "spark_catalog")
      .getOrCreate()
```

**Comment by braislchao** [comment]
To make the iceberg integration work, we needed to add the iceberg libraries as _--extra-jars_. 
This worked maintaining the _--datalake-formats_ property as delta:

```
"--extra-jars" = "s3://factorial-metabolic/extra-libs/metabolic-core-iceberg-assembly.jar,s3://factorial-metabolic/extra-libs/iceberg-aws-bundle-1.6.1.jar,s3://factorial-metabolic/extra-libs/iceberg-spark-runtime-3.3_2.12-1.6.1.jar"
"--datalake-formats" = "delta"
```

In the previous configuration we received an error of missing warehouse location. Seems like the warehouse must be explicitly for glue to be able to write in a s3 folder as the main catalog: 

> org.apache.iceberg.exceptions.ValidationException: Cannot derive default warehouse location, warehouse path must not be null or empty at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:49)

This property must be configured in Metabolic Spark configuration:

```
.config("spark.sql.catalog.spark_catalog.warehouse", "s3://factorial-metabolic/data-lake/dev/feature_test_iceberg_write/")
```

**Comment by braislchao** [comment]
Added two changes not related with this PR but needed for test integrity:
- Change `log4j2.properties` configuration in test environment
- Ignore Secret Manager test  (probably we need to solve the Token issues but first I need to understand the functionality of this test)

File: build.sbt [hunk 1]
```diff
@@ -71,7 +74,7 @@ libraryDependencies += "mysql" % "mysql-connector-java" % "5.1.49" % "test"
 
 libraryDependencies += "org.mockito" % "mockito-core" % "2.21.0" % "test"
 
-Test / fork := true
+//Test / fork := true
```
Comment (margon8) [inline build.sbt:77]: why?

File: src/main/scala/com/metabolic/data/core/services/spark/writer/file/IcebergWriter.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        .saveAsTable(output_identifier)
+
+      case WriteMode.Overwrite => df
+        .write
+        .format("iceberg")
+        .mode("overwrite")
+        .saveAsTable(output_identifier)
+
+      case WriteMode.Upsert =>
+        throw new NotImplementedError("Batch Upsert is not supported in Iceberg yet")
```
Comment (margon8) [inline src/main/scala/com/metabolic/data/core/services/spark/writer/file/IcebergWriter.scala:38]: How would they do incremental overwrites?

File: src/main/scala/com/metabolic/data/mapper/app/MetabolicApp.scala [hunk 3]
```diff
@@ -33,11 +37,21 @@ class MetabolicApp(sparkBuilder: SparkSession.Builder) extends Logging {
...
       .config("spark.databricks.delta.schema.autoMerge.enabled", "true")
       .config("spark.databricks.delta.optimize.repartition.enabled", "true")
       .config("spark.databricks.delta.vacuum.parallelDelete.enabled", "true")
+      .config("spark.sql.catalog.prod", "org.apache.iceberg.spark.SparkSessionCatalog")
+      .config("spark.sql.catalog.prod.catalog-impl", "org.apache.iceberg.aws.glue.GlueCatalog")
+      .config("spark.sql.catalog.prod.io-impl", "org.apache.iceberg.aws.s3.S3FileIO")
+      .config("spark.sql.catalog.prod.client.region", s"$client_region")
+      .config("spark.sql.catalog.prod.warehouse", s"$warehouse_global")
+      .config("spark.sql.catalog.spark_catalog", "org.apache.iceberg.spark.SparkSessionCatalog")
+      .config("spark.sql.catalog.spark_catalog.catalog-impl", "org.apache.iceberg.aws.glue.GlueCatalog")
```
Comment (margon8) [inline src/main/scala/com/metabolic/data/mapper/app/MetabolicApp.scala:50]: Shouldn't this be a param for non glue users?

Comment (braislchao) [reply src/main/scala/com/metabolic/data/mapper/app/MetabolicApp.scala:50]: I think so, maybe we can use a default spark.conf and let users customize everything

File: src/main/scala/com/metabolic/data/core/services/spark/writer/file/IcebergWriter.scala (hunk 2 above)
Comment (braislchao) [reply src/main/scala/com/metabolic/data/core/services/spark/writer/file/IcebergWriter.scala:38]: We definitely need to add the MERGE INTO option

Comment (braislchao) [reply src/main/scala/com/metabolic/data/core/services/spark/writer/file/IcebergWriter.scala]: Comming back to this, incremental overwrite should be done with the op replace, which replaces the table with the information of the dataframe regardless of the schema. 

I have added an implementation of the op overwrite which uses https://iceberg.apache.org/docs/1.6.0/spark-writes/#overwriting-data . This requires the new data to maintain the same schema,

**Comment by braislchao** [comment]
Upsert working with Glue Catalog, added tests.

**Review by braislchao — APPROVED** [review]
Merging after testing several mappings

**Pull request merged**